applications/tari_dan_wallet_daemon/src/handlers/wasm_optimizer.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2025 The Tari Project
+// SPDX-License-Identifier: BSD-3-Clause
+
+use std::io;
+
+use tempfile::tempdir;
+use thiserror::Error;
+use tokio::{fs::File, io::AsyncWriteExt};
+use wasm_opt::{OptimizationError, OptimizationOptions};
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("I/O error: {0}")]
+    IO(#[from] io::Error),
+    #[error("Optimization error: {0}")]
+    Optimization(#[from] OptimizationError),
+    #[error("Invalid file after optimization: {0}")]
+    InvalidFile(String),
+}
+
+/// Optimizes a WebAssembly (WASM) template binary to reduce its size.
+///
+/// # Arguments
+///
+/// * `template_binary` - The original WASM binary to optimize
+///
+/// # Returns
+///
+/// A `Result` containing either the optimized WASM binary as a `Vec<u8>` or an `Error`
+///
+/// # Errors
+///
+/// Returns an error if:
+/// - There are I/O errors when creating temporary files
+/// - The optimization process fails
+pub struct WasmTemplateOptimizer {}
+
+impl WasmTemplateOptimizer {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    pub async fn optimize(&self, template_binary: &[u8]) -> Result<Vec<u8>, Error> {
+        let temp_dir = tempdir()?;
+
+        // create temporary input file
+        let input_file_path = temp_dir.path().join("input.wasm");
+        let output_file_path = temp_dir.path().join("output.wasm");
+        let mut input_file = File::create(input_file_path.as_path()).await?;
+        input_file.write_all(template_binary).await?;
+        input_file.flush().await?;
```

Comment:
input_file will only close at the end of the function
```suggestion
        tokio::fs::write(&input_file_path, template_binary).await?;
```

Reply:
true, its better thanks!